statbot/crawler.py

```diff
@@ -0,0 +1,194 @@
+from .range import Range
+from .util import null_logger
+
+__all__ = [
+    'DiscordHistoryCrawler',
+]
+
+NOW_ID = discord.utils.time_snowflake(datetime.now())
+
+class DiscordHistoryCrawler:
+    __slots__ = (
+        'client',
+        'sql',
+        'config',
+        'logger',
+        'channels',
+        'finished',
+        'progress',
+        'queue',
+    )
+
+    def __init__(self, client, sql, config, logger=null_logger):
+        self.client = client
+        self.sql = sql
+        self.config = config
+        self.logger = logger
+        self.channels = {} # {channel_id : channel}
+        self.progress = {} # {channel_id : MessageHistory}
+        self.queue = asyncio.Queue(self.config['crawler']['queue-size'])
+
+    async def _init_channels(self):
+        async with self.sql.orm.transaction():
+            for guild in self.client.guilds:
+                if guild.id in self.config['guilds']:
+                    for channel in guild.text_channels:
+                        if channel.permissions_for(guild.me).read_message_history:
+                            self.channels[channel.id] = channel
+                            mhist = self.sql.orm.lookup_message_hist(channel)
+
+                            if mhist is None:
+                                mhist = MessageHistory()
+                                self.sql.orm.insert_message_hist(channel, mhist)
+
+                            self.progress[channel.id] = mhist
+
+        # Remove deleted channels from tracker
+        for channel in set(self.progress.keys()) - set(self.channels.keys()):
+            del self.progress[channel.id]
+
+    def start(self):
+        self.client.hooks['on_guild_channel_create'] = self._channel_create_hook
+        self.client.hooks['on_guild_channel_delete'] = self._channel_delete_hook
+        self.client.hooks['on_guild_channel_update'] = self._channel_update_hook
+
+        self.client.loop.create_task(self.producer())
+        self.client.loop.create_task(self.consumer())
+
+    async def producer(self):
+        self.logger.info("Producer coroutine started!")
+
+        # Setup
+        await self.client.wait_until_ready()
+        await self._init_channels()
+
+        yield_delay = self.config['crawler']['yield-delay']
+        long_delay = self.config['crawler']['long-delay']
+
+        while True:
+            # tuple() is necessary since the underlying
+            # dict of channels may change size during
+            # an iteration
+            all_empty = True
+            for cid in tuple(self.progress.keys()):
+                # Do round-robin between all the channels
+                try:
+                    channel = self.channels[cid]
+                    mhist = self.progress[cid]
+                    all_empty &= not await self._read(channel, mhist)
+                except BaseException as ex:
+                    if isinstance(ex, SystemExit):
+                        raise ex
```

Comment:
wait, can we not do this? this also catches `SystemExit`, `KeyboardInterrupt`, `GeneratorExit`

let's just catch `Exception` and not reraise (also, if we're going to reraise, use `raise` instead of `raise ex` to preserve the stack trace)